### DISCUSSION TITLE
汇编魔法实现 C++ 协程 | SF-Zhou's Blog

### DISCUSSION THREAD

**Discussion by SF-Zhou** [SF-Zhou/sf-zhou.github.io#164] [Announcements]
https://sf-zhou.github.io/programming/cpp_magic_coroutine.html

**Comment by SF-Zhou** [comment]
PS：`libgo` 提供了一个开启协程的语法糖：
```c++
go [] {
  std::cout << "Run" << std::endl;
};
```
第一眼看上去觉得这不可能是 C++。然而这里的 `go` 其实是一个宏，类似对应 `Go()-`，然后 `Go` 类里重载减号运算符，相当精妙了。

**Comment by SF-Zhou** [comment]
百度的 [brpc](https://github.com/apache/incubator-brpc) 项目中也有自研的 bthread M:N 线程库，与 libgo 的实现类似，详情可以参看其[文档](https://github.com/apache/incubator-brpc/blob/master/docs/cn/bthread.md)。

**Comment by pedrogao** [comment]
深度好文，但是我在学习的时候，发现例子无法运行，全都是`段错误`啊

**Comment by pedrogao** [comment]
这里应该是 g++ 编译优化导致访问内存越界，造成的段错误。
我使用如下方式则运行正常：
```sh
g++ -S -Og swap.cpp
g++ -c swap.s -o swap.o
g++ swap.o -o swap
```
先将代码编译成汇编，防止 g++ 对汇编代码优化。

直接解决问题的方式就是给`-O`参数指定优化等级，如`g`等级：
```sh
g++ -Og swap.cpp
```

**Reply by Yikai-Liao** [reply]
你好，我在MAC 和 Linux上进行了测试，MAC上是完全正常的。
但是在Linux 上，如果在Clion中使用CMake的Debug配置文件就会报`段错误`，反而优化等级更高的 Release 配置文件就可以正常运行。
我尝试直接在命令行用 g++ 进行编译，发现各个优化等级都是可以正常运行的

**Reply by Yikai-Liao** [reply]
直接不开优化进行编译反而无法运行 例如
```bash
g++ main.cpp
```

**Comment by pedrogao** [comment]
麻烦您解答一下，关于这个地方的内存对齐处理的原理是什么：
```c
eg[5] = (char *)((uintptr_t)(&mem.back()) & ~15ull) - sizeof(void *);
```
多谢了～

**Comment by SF-Zhou** [comment]
> @PedroGao
> 麻烦您解答一下，关于这个地方的内存对齐处理的原理是什么：
> ```c
> eg[5] = (char *)((uintptr_t)(&mem.back()) & ~15ull) - sizeof(void *);
> ```
> 多谢了～

我更新了文中的在线执行链接，加上 `-O2` 现在执行没问题了。
这里对齐是将栈底指针强制对齐为 `stack % 16 == 8`，是 x86 ABI 的要求了。可以参考 https://stackoverflow.com/questions/26866723/main-and-stack-alignment

**Comment by pedrogao** [comment]
cool~, I got it. 有对栈对齐感兴趣的同学可以查阅：https://www.cnblogs.com/tcctw/p/11333743.html。

**Comment by SF-Zhou** [comment]
> @PedroGao
> cool~, I got it. 有对栈对齐感兴趣的同学可以查阅：https://www.cnblogs.com/tcctw/p/11333743.html。

补充一点，内存地址对齐也是数据一致性的要求。CPU 原子操作指令所操作的数据地址不能跨过两个 CacheLine，而对齐可以保证这一点。`struct` 结构体中的也有类似的对齐策略。

**Comment by hecyxy** [comment]
你好，请教个问题：(char *)((uintptr_t)(&mem.back()) & ~15ull) - sizeof(void *)，为啥要减去sizeof(void *)呀

**Reply by SF-Zhou** [reply]
[看这里的解释](https://github.com/SF-Zhou/sf-zhou.github.io/discussions/164#discussioncomment-2070392)

**Reply by hecyxy** [reply]
多谢！么注意到。

**Comment by Hotwill** [comment]
```ams
_Z4funcv:
  push rbp     ; 存储 RBP 的值
  mov rbp, rsp ; 将 RPB 指向当前栈顶 RSP
  sub rsp, 12  ; 将栈顶上移 12 个字节，为临时变量 a/b/c 腾出空间
               ; a = [rbp - 4], b = [rbp - 8], c = [rbp - 12]
  mov rsp, rbp ; 恢复栈顶指针 RSP
  pop rbp      ; 恢复 RBP 的值，即为函数调用者的栈帧地址
  ret          ; 读取栈顶存储的返回地址，弹出，返回

main:
  push rbp
  mov rbp, rsp
  call _Z4funcv
  mov eax, 0
  pop rbp
  ret
```
这里的
```
  push rbp     ;
  mov rbp, rsp ;
```
是不是重复了？

**Reply by SF-Zhou** [reply]
没有重复，可以参考 https://godbolt.org/z/4Kr1c9Tox 这里汇编的内容
